File: src/main/java/work/lclpnet/ap2/game/apocalypse_survival/util/PursuitClass.java
package work.lclpnet.ap2.game.apocalypse_survival.util;

import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.mob.MobEntity;
import net.minecraft.server.network.ServerPlayerEntity;
import org.jetbrains.annotations.Nullable;
import work.lclpnet.ap2.api.base.Participants;

import java.util.*;
import java.util.function.Supplier;

public class PursuitClass<T extends MobEntity> {

    private final Participants participants;
    private final Map<UUID, Pursuit<T>> pursuitMap;
    private final Set<T> mobs = new HashSet<>();
    private final Map<T, Pursuit<T>> pursuitMember = new HashMap<>();
    private final TargetManager debug$targetManager;

    public PursuitClass(Participants participants, int capacityPerPlayer, TargetManager debug$targetManager) {
        this.participants = participants;
        this.pursuitMap = new HashMap<>(participants.count());
        this.debug$targetManager = debug$targetManager;

        for (ServerPlayerEntity player : participants) {
            UUID uuid = player.getUuid();

            Supplier<@Nullable LivingEntity> playerGetter = () -> participants.getParticipant(uuid).orElse(null);

            var pursuit = new Pursuit<T>(playerGetter, capacityPerPlayer);

            pursuitMap.put(uuid, pursuit);
        }
    }

    public void addMob(T mob) {
        mobs.add(mob);
    }

    public void removeMob(T mob) {
        mobs.remove(mob);
        stopPursuit(mob);
    }

    public void removeParticipant(ServerPlayerEntity player) {
        pursuitMap.remove(player.getUuid());
    }

    public void update() {
        // set all pursuit instances to need an update, because mobs will have moved since the last update
        for (var pursuit : pursuitMap.values()) {
            pursuit.setNeedsUpdate();
        }

        // update each mobs target
        nextMob: for (T mob : mobs) {
            for (ServerPlayerEntity player : playersByDistance(mob)) {
                var pursuit = pursuitMap.get(player.getUuid());

                if (pursuit == null) continue;

                if (pursuit.hasCapacity()) {
                    changePursuit(mob, player, pursuit);
                    continue nextMob;
                }

                // check if the current mob can replace the most distant pursuer because it is closer to the player
                T mostDistant = pursuit.getMostDistantPursuer();

                if (mostDistant == null || mob.squaredDistanceTo(player) >= mostDistant.squaredDistanceTo(player)) {
                    continue;
                }

                // current mob is closer, replace old pursuer.
                // the old pursuer will only be updated in the next iteration.
                // this could theoretically be changed to another pass in this update
                stopPursuit(mostDistant);
                changePursuit(mob, player, pursuit);
            }

            // didn't find a player to pursuit, start roaming (RoamGoal will pick up)
        }
    }

    private Iterable<ServerPlayerEntity> playersByDistance(T mob) {
        return () -> participants.stream()
                .sorted(Comparator.comparingDouble(mob::squaredDistanceTo))
                .iterator();
    }

    private void changePursuit(T mob, LivingEntity target, Pursuit<T> pursuit) {
        var oldPursuit = pursuitMember.put(mob, pursuit);

        if (oldPursuit != null) {
            oldPursuit.removePursuer(mob);
        }

        if (pursuit.addPursuer(mob)) {
            mob.setTarget(target);

            debug$targetManager.debug$getScoreboardManager().joinTeam(mob, debug$targetManager.debug$getPursuitTeam());
        } else {
            stopPursuit(mob);
        }
    }

    private void stopPursuit(T mob) {
        var pursuit = pursuitMember.remove(mob);

        if (pursuit != null) {
            pursuit.removePursuer(mob);
        }

        mob.setTarget(null);

        debug$targetManager.debug$getScoreboardManager().leaveTeam(mob, debug$targetManager.debug$getPursuitTeam());
    }
}

File: src/minigames/apocalypse_survival/src/main/kotlin/work/lclpnet/ap2/game/apocalypse_survival/util/PursuitClass.kt
package work.lclpnet.ap2.game.apocalypse_survival.util

import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.entity.Mob
import work.lclpnet.ap2.api.base.Participants
import java.util.UUID

private const val INITIAL_FORCED_TARGET_SECONDS = 10

class PursuitClass<T : Mob>(
    private val participants: Participants,
    private val capacityPerPlayer: Int
) {

    private val pursuitMap = HashMap<UUID, Pursuit<T>>(participants.count())
    private val mobs = HashSet<T>()
    private val pursuitMember = HashMap<T, Pursuit<T>>()
    private val forcedTarget = Object2IntOpenHashMap<T>()

    init {
        for (player in participants) {
            val uuid = player.uuid
            val playerGetter: () -> LivingEntity? = { participants.getParticipant(uuid).orElse(null) }
            pursuitMap[uuid] = Pursuit(playerGetter, capacityPerPlayer)
        }
    }

    fun addMob(mob: T) {
        mobs.add(mob)
        forceTargetInitially(mob)
    }

    fun removeMob(mob: T) {
        mobs.remove(mob)
        stopPursuit(mob)
    }

    fun removeParticipant(player: ServerPlayer) {
        pursuitMap.remove(player.uuid)
    }

    private fun forceTargetInitially(mob: T) {
        if (mobs.size >= participants.count() * capacityPerPlayer) return

        val player = playerWithLeastPursuers() ?: return
        val pursuit = pursuitMap[player.uuid] ?: return

        forcedTarget.put(mob, INITIAL_FORCED_TARGET_SECONDS)
        changePursuit(mob, player, pursuit)
    }

    fun update() {
        for (pursuit in pursuitMap.values) {
            pursuit.setNeedsUpdate()
        }

        nextMob@ for (mob in mobs) {
            if (forcedTarget.containsKey(mob) && shouldKeepForcedTarget(mob)) continue

            for (player in playersByDistance(mob)) {
                val pursuit = pursuitMap[player.uuid] ?: continue

                if (pursuit.hasCapacity()) {
                    changePursuit(mob, player, pursuit)
                    continue@nextMob
                }

                val mostDistant = pursuit.getMostDistantPursuer()

                if (mostDistant == null || mob.distanceToSqr(player) >= mostDistant.distanceToSqr(player)) {
                    continue
                }

                stopPursuit(mostDistant)
                changePursuit(mob, player, pursuit)
            }
        }
    }

    private fun shouldKeepForcedTarget(mob: T): Boolean {
        val remaining = forcedTarget.getInt(mob)

        if (remaining <= 1) {
            forcedTarget.removeInt(mob)
            return false
        }

        forcedTarget.put(mob, remaining - 1)
        return true
    }

    private fun playersByDistance(mob: T): Iterable<ServerPlayer> =
        Iterable { participants.stream().sorted(Comparator.comparingDouble { mob.distanceToSqr(it) }).iterator() }

    private fun changePursuit(mob: T, target: LivingEntity, pursuit: Pursuit<T>) {
        val oldPursuit = pursuitMember.put(mob, pursuit)
        oldPursuit?.removePursuer(mob)

        if (pursuit.addPursuer(mob)) {
            mob.target = target
        } else {
            stopPursuit(mob)
        }
    }

    private fun stopPursuit(mob: T) {
        pursuitMember.remove(mob)?.removePursuer(mob)
        mob.target = null
        forcedTarget.removeInt(mob)
    }

    private fun playerWithLeastPursuers(): ServerPlayer? =
        participants.stream().min(Comparator.comparingInt { player ->
            pursuitMap[player.uuid]?.getPursuerCount() ?: Int.MAX_VALUE
        }).orElse(null)
}
